cmd/main.go

```diff
@@ -54,7 +65,110 @@ func main() {
 		metricsServer := agent.NewMetricsServer(ambAgent.MetricsRelayHandler)
 		return metricsServer.Serve(ctx, metricsListener)
 	})
-	grp.Go("watch", ambAgent.Watch)
+
+	// Begin lease-lock. watch when we are leader
+	grp.Go("lease-lock-watch", func(ctx context.Context) error {
+		// use a Go context so we can tell the leaderelection code when we
+		// want to step down
+		ctx, cancel := context.WithCancel(ctx)
+		defer cancel()
+
+		// listen for interrupts or the Linux SIGTERM signal and cancel
+		// our context, which the leader election code will observe and
+		// step down
+		ch := make(chan os.Signal, 1)
+		signal.Notify(ch, os.Interrupt, unix.SIGTERM)
+		go func() {
+			<-ch
+			dlog.Info(ctx, "Received termination, signaling shutdown")
+			cancel()
+			os.Exit(0)
+		}()
+
+		// each call to the leaselock should have a unique id
+		id := uuid.New().String()
+		dlog.Infof(ctx, "Will lease with id %s", id)
+
+		// we use the Lease lock type since edits to Leases are less common
+		// and fewer objects in the cluster watch "all Leases".
+		lock := &resourcelock.LeaseLock{
+			LeaseMeta: metav1.ObjectMeta{
+				Name:      "ambassador-agent-lease-lock",
+				Namespace: ambAgent.AgentNamespace,
+			},
+			Client: clientset.CoordinationV1(),
+			LockConfig: resourcelock.ResourceLockConfig{
+				Identity: id,
+			},
+		}
+
+		// check if we have permissions
+		_, _, err = lock.Get(ctx)
+		if err != nil {
+			se := &apierrors.StatusError{}
+			if errors.As(err, &se) && se.Status().Code == http.StatusForbidden {
+				// if we do not have permissions, skip leader election
+				dlog.Warnf(ctx, "%s%s",
+					"Agent has no permissions to work with leases; will disable leader election. This may be inefficient. ",
+					"To fix, please install the agent from a new version of its helm chart.",
+				)
```

Comment:
what's up with this? if anything, you can also just do `dlog.Warn(ctx, "foo", "bar", "baz")` and skip the formatting